ENH: Improve error message for header argument containing non int types. GH16338

mjlove12 opened the pull request:
Adds error "header must be integer or list of integers" when the header argument is a list, tuple or numpy array containing non-integers. Initially intended to read_csv, but applies to other functions with similar header arguments. GH16338 refers to a case in which the user mixes up the "names" and "header" arguments.

 - [x] closes #16338
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [x] whatsnew entry

mjlove12 committed:
ENH: Improve error message for read_csv header argument containing non int types. GH16338.

Adds error "header must be integer or list of integers" when the header argument is a list, tuple or numpy array containing non-integers. Initially intended to read_csv, but applies to other functions with similar header arguments. GH16338 refers to a case in which the user mixes up the "names" and "header" arguments.

Revising PR16351 based on feedback

Revising PR16351 lint issues

Adding release note in whatsnew v0.21.0 for PR16351

TomAugspurger commented:
Lint failure [here](https://travis-ci.org/pandas-dev/pandas/jobs/232024816#L1398) If you `pip install flake8` you can run `git diff upstream/master --name-only -- '*.py' | flake8 --diff` locally.

The other failure looked unrelated, so I've restarted that job. Once you've fixed the linting error, you can push another commit to this branch.

gfyoung commented:
Also need to test that you can hit this error message!

pierre-haessig commented:
Thanks @mjlove12 for picking this up.

The lint error is `pandas/io/parsers.py:1188:1: W293 blank line contains whitespace`. It will be fixed by removing the whitespace on line 1188 (which is a blank line).

mjlove12 commented:
Thanks @pierre-haessig. I'm planning to make the changes this week,

mjlove12 commented:
Okay, I hope I did this right.
1. Added extra lines to account for cases when the header is non-array-like. 
2. Added a test (test_non_int_header()) in the header.py file under the Header_Tests class which includes a non-array-like case and an array-like case.
3. Added changes with "git add ."
4. Committed the changes to my branch with "git commit -m 'Revising PR16351 based on feedback'"
5. Rebased with "git rebase -i origin/master" and squashing the second commit instead of "picking" it
6. Forced a push to origin with "git push origin new-read_csv-header-comments -f"

Feedback appreciated if any of those steps are discouraged. 

Also, random question:
Can anyone explain why the test cases are wrapped in classes and how that works? I tried to match some of the other tests, but didn't understand why we were using the test classes to call methods like "self.read_csv" and "self.read_table" instead of just pd.read_csv. 

Thanks for all the help!
Matt

gfyoung commented:
@mjlove12 : Your steps are perfectly fine.  Don't worry about it too much. :smile:

As for why we use `self.read_csv`, it's because there are multiple engines that can be used when parsing a CSV.  The two main ones are:
* C engine: this is `pandas` in-house parser backed up in C and Cython
* Python engine: this leverages Python's `csv` library

mjlove12 commented:
@TomAugspurger not entirely sure I'm understanding the failure this time. I ran the command "git diff master --name-only -- '*.py' | grep 'pandas/' | xargs flake8" on the branch and didn't get any issues. Any help would be appreciated.

TomAugspurger commented:
@mjlove12 [this test failure](https://travis-ci.org/pandas-dev/pandas/jobs/234430487) looks unrelated. I've restarted that worker.

TomAugspurger commented:
You can also add a release note in `doc/source/whatsnew/v0.21.0.txt` if you want, under the bug fixes section.

mjlove12 commented:
@TomAugspurger sounds good. I added a whatsnew entry under bug fixes.

TomAugspurger commented:
The appveyor failure is unrelated, and already fixed on master.

Thanks @mjlove12!
